File: app/src/main/java/uk/co/ribot/androidboilerplate/data/local/DbOpenHelper.java
package uk.co.ribot.androidboilerplate.data.local;

import android.arch.persistence.db.SupportSQLiteDatabase;
import android.arch.persistence.db.SupportSQLiteOpenHelper;
import android.arch.persistence.db.framework.FrameworkSQLiteOpenHelperFactory;
import android.content.Context;

import com.sunny.common.injection.qualifier.ApplicationContext;
import com.sunny.common.utils.LogUtil;
import com.sunny.datalayer.model.bean.Subject;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class DbOpenHelper {

    public static final String DATABASE_NAME = "ribots.db";
    public static final int DATABASE_VERSION = 2;

    private SupportSQLiteOpenHelper mOpenHelper;

    @Inject
    public DbOpenHelper(@ApplicationContext Context context) {
        SupportSQLiteOpenHelper.Configuration configuration = SupportSQLiteOpenHelper.Configuration
                .builder(context)
                .name(DATABASE_NAME)
                .callback(new DbCallback(DATABASE_VERSION))
                .build();
        FrameworkSQLiteOpenHelperFactory factory = new FrameworkSQLiteOpenHelperFactory();
        mOpenHelper = factory.create(configuration);
    }

    public SupportSQLiteOpenHelper getOpenHelper() {
        return mOpenHelper;
    }

    private static class DbCallback extends SupportSQLiteOpenHelper.Callback {
        DbCallback(int version) {
            super(version);
        }

        @Override
        public void onConfigure(SupportSQLiteDatabase db) {
            super.onConfigure(db);
            //Uncomment line below if you want to enable foreign keys
            //db.execSQL("PRAGMA foreign_keys=ON;");
        }

        @Override
        public void onCreate(SupportSQLiteDatabase db) {
            db.beginTransaction();
            try {
                db.execSQL(Subject.CREATE_TABLE);
                //Add other tables here
                db.setTransactionSuccessful();
            } finally {
                db.endTransaction();
            }
        }

        @Override
        public void onUpgrade(SupportSQLiteDatabase db, int oldVersion, int newVersion) {
            LogUtil.i("升级数据库：db = [" + db.getPath() + "], " +
                    "oldVersion = [" + oldVersion + "], currentVersion = [" + newVersion + "]");
        }
    }

}

File: app/src/main/java/uk/co/ribot/androidboilerplate/data/local/DbOpenHelper.kt
package uk.co.ribot.androidboilerplate.data.local

import android.arch.persistence.db.SupportSQLiteDatabase
import android.arch.persistence.db.SupportSQLiteOpenHelper
import android.arch.persistence.db.framework.FrameworkSQLiteOpenHelperFactory
import android.content.Context
import uk.co.ribot.androidboilerplate.data.model.bean.Subject
import uk.co.ribot.androidboilerplate.injection.qualifier.ApplicationContext
import uk.co.ribot.androidboilerplate.utils.i
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DbOpenHelper @Inject constructor(@ApplicationContext context: Context) {

    companion object {
        const val DATABASE_NAME = "ribots.db"
        const val DATABASE_VERSION = 2
    }

    val openHelper: SupportSQLiteOpenHelper

    init {
        val configuration = SupportSQLiteOpenHelper.Configuration
                .builder(context)
                .name(DATABASE_NAME)
                .callback(DbCallback(DATABASE_VERSION))
                .build()
        val factory = FrameworkSQLiteOpenHelperFactory()
        openHelper = factory.create(configuration)
    }

    private class DbCallback constructor(version: Int) : SupportSQLiteOpenHelper.Callback(version) {
        override fun onConfigure(db: SupportSQLiteDatabase) {
            super.onConfigure(db)
            //Uncomment line below if you want to enable foreign keys
            //db.execSQL("PRAGMA foreign_keys=ON;");
        }

        override fun onCreate(db: SupportSQLiteDatabase) {
            db.beginTransaction()
            try {
                db.execSQL(Subject.CREATE_TABLE)
                //Add other tables here
                db.setTransactionSuccessful()
            } finally {
                db.endTransaction()
            }
        }

        override fun onUpgrade(db: SupportSQLiteDatabase, oldVersion: Int, newVersion: Int) {
            i("升级数据库：db = [" + db.path + "], " +
                    "oldVersion = [" + oldVersion + "], currentVersion = [" + newVersion + "]")
        }
    }

}
